pytradfri/api/aiocoap_api.py

```diff
@@ -218,21 +232,26 @@ async def _observe(self, api_command, timeout):
 
         api_command.result = _process_output(pr_rsp)
 
-        def success_callback(res):
+        def success_callback(res: Message) -> None:
             api_command.result = _process_output(res)
 
-        def error_callback(exc):
+        def error_callback(exc: Exception) -> None:
             if isinstance(exc, LibraryShutdown):
                 _LOGGER.debug("Protocol is shutdown, stopping observation")
                 return
-            err_callback(exc)
+
+            if err_callback:
+                err_callback(exc)
 
         observation = pr_req.observation
+        # The observation is set on the request
+        # since we pass a Message with observe set above.
+        assert observation is not None
```

Comment:
I'm using Bandit for another project (and I've seen core uses this too). It throws a warning when using `assert` so should we consider raising instead?
https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html

Reply:
We use assert here to tell mypy that we know that the observation is not None. It's not used during runtime.

Reply:
Having “assert” in production code in not a good idea…it is not used for the runtime logic, but the assert is executed.

Reply:
It's executed if it's not turned off, yes. We can check and raise some error instead. I just want to make clear that we don't do this for any runtime purposes. It's just to satisfy the static check of mypy.

If we would be wrong about our assumptions it's a bug and we would raise anyway below.